packages/react/src/lazy.ts

```diff
@@ -108,3 +109,126 @@ const createLazy =
 export const lazy = Object.assign(createLazy({}), {
   create: createLazy,
 })
+
+export interface ReloadOnErrorStorage {
+  getItem: (key: string) => string | null
+  setItem: (key: string, value: string) => void
+  removeItem: (key: string) => void
+}
+
+export interface ReloadOnErrorOptions extends LazyOptions {
+  /**
+   * The number of times to retry the loading of the component. \
+   * If `true`, the component will be retried indefinitely.
+   *
+   * @default 1
+   */
+  retry?: number | boolean
+  /**
+   * The delay between retries. \
+   * If a function is provided, it will be called with the current retry count.
+   *
+   * @default 0
+   */
+  delay?: number | ((retryCount: number) => number)
+  /**
+   * The storage to use for the retry count. \
+   * If not provided, it assumes that you are in a browser environment and uses `sessionStorage`.
+   */
+  storage?: ReloadOnErrorStorage
+  /**
+   * The function to use to reload the component. \
+   * If not provided, it assumes that you are in a browser environment and uses `window.location.reload`.
+   */
+  reload?: () => void
+}
+
+export const reloadOnError = ({ retry = 1, delay = 0, storage, reload, ...options }: ReloadOnErrorOptions) => {
+  const getDefaultStorage = (): ReloadOnErrorStorage | null => {
+    if (storage) {
+      return storage
+    }
+
+    if (typeof window === 'undefined' && 'sessionStorage' in window) {
+      return null
+    }
+
+    return window.sessionStorage
+  }
+
+  const getDefaultReloadFunction = (): ReloadOnErrorOptions['reload'] | null => {
+    if (reload) {
+      return reload
+    }
+
+    if (typeof window === 'undefined' && 'location' in window) {
+      return null
+    }
+
+    return () => window.location.reload()
+  }
+
+  const reloadStorage = getDefaultStorage()
+  const reloadFunction = getDefaultReloadFunction()
+
+  /**
+   * Parses the reload count from the storage.
+   */
+  const parseReload = (value: string | null): { reloadCount: number; isNaN: boolean } => {
+    if (!value) {
+      return {
+        reloadCount: 0,
+        isNaN: true,
+      }
+    }
+
+    const reloadCount = parseInt(value, 10)
+    const isNaN = Number.isNaN(reloadCount)
+
+    return {
+      reloadCount: isNaN ? 0 : reloadCount,
+      isNaN,
+    }
+  }
+
+  return createLazy({
+    ...options,
+    onSuccess: ({ load }) => {
+      options.onSuccess?.({ load })
+      reloadStorage?.removeItem(load.toString())
+    },
+    onError: ({ error, load }) => {
+      options.onError?.({ error, load })
+
+      if (reloadStorage == null || reloadFunction == null) {
+        return
+      }
```

Comment:
reloadStroage, reloadFunction could be Nonnullable if throw error when there is no reloadStorage, reloadFunction